src/qdisp/QueryRequest.cc

```diff
@@ -407,74 +448,103 @@ void QueryRequest::ProcessResponseData(XrdSsiErrInfo const& eInfo,
-                                      << "dataIgnored=" << dataIgnored);
-            }
-        }
+    if (jq->isQueryCancelled()) {
+        LOGS(_log, LOG_LVL_WARN, "QueryRequest::_processData job was cancelled.");
         _errorFinish(true);
         return;
     }
 
     // Get a copy of the shared buffer pointer so askForResponseDataCmd can be deleted.
     ResponseHandler::BufPtr bufPtr = _askForResponseDataCmd->getBufPtr();
     _askForResponseDataCmd.reset(); // No longer need it, and don't want the destructor calling _errorFinish().
+
+    int const protoHeaderSize = proto::ProtoHeaderWrap::getProtoHeaderSize();
+    ResponseHandler::BufPtr nextHeaderBufPtr;
+
+    // Values for these variables to be filled in by flush() calls.
     bool largeResult = false;
     int nextBufSize = 0;
     int resultRows = 0;
-    bool flushOk = jq->getDescription()->respHandler()->flush(blen, bufPtr, last,
-                                                              largeResult, nextBufSize, resultRows);
+    bool last = false;
+    bool flushOk = false;
+
+    // The buffer has 2 parts.
+    // - The first (bytes = blen - ProtoHeaderWrap::getProtheaderSize())
+    //   is the result associated with the previously received header.
+    // - The second is the header for the next message.
+    int respSize = blen - protoHeaderSize;
+    nextHeaderBufPtr = make_shared<vector<char>>(bufPtr->begin() + respSize, bufPtr->end());
+    // Read the result
+    // Values for last, largeResult, and nextBufSize filled in by flush
+    flushOk = jq->getRespHandler()->flush(respSize, bufPtr, last, largeResult, nextBufSize, resultRows);
+    if (last) {
+        // Last should only be true when the header is read, not the result.
+        throw Bug("_processData result had 'last' true, which cannot be allowed.");
+    }
+
+    _totalRows += resultRows;
+
+    bufPtr.reset(); // don't need the buffer anymore and it could be big.
+    if (nextBufSize != protoHeaderSize) {
+        throw Bug("Unexpected header size from flush(result) call QID="
+                + to_string(_qid) + "#" + to_string(_jobid));
+    }
+    if (!flushOk) {
+        _flushError(jq);
+        return;
+    }
+
+    // Read the next header
+    // Values for last, largeResult, and nextBufSize filled in by flush
+    // resultRows ignored in headers.
+    flushOk = jq->getRespHandler()->flush(protoHeaderSize, nextHeaderBufPtr, last, largeResult, nextBufSize, resultRows);
+
     if (largeResult) {
         if (!_largeResult) LOGS(_log, LOG_LVL_DEBUG, "holdState largeResult set to true");
         _largeResult = true; // Once the worker indicates it's a large result, it stays that way.
     }
 
     if (flushOk) {
-        _totalRows += resultRows;
+        if (last != xrdLast) {
+            LOGS(_log, LOG_LVL_DEBUG, "processData disagreement between last=" << last
+                                     << " and xrdLast=" << xrdLast);
```

Comment:
xrdLast is inconsistent, while last is authoratative/correct. This is just a footnote in case it becomes important to figure out why xrootd disagrees on last. Not really a happy fun thing, but it has been around for many years.